crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -114,30 +423,123 @@ impl<S: States> Deref for State<S> {
     }
 }
 
+#[derive(Resource, Default, Debug)]
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(bevy_reflect::Reflect),
+    reflect(Resource, Default)
+)]
+struct PreviousState<S: States>(S);
+
+impl<S: States> PreviousState<S> {
+    /// Get the current state.
+    pub fn get(&self) -> &S {
+        &self.0
+    }
+}
+
+impl<S: States + PartialEq> PartialEq<S> for PreviousState<S> {
+    fn eq(&self, other: &S) -> bool {
+        self.get() == other
+    }
+}
+
+impl<S: States> Deref for PreviousState<S> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        self.get()
+    }
+}
+
 /// The next state of [`State<S>`].
 ///
 /// To queue a transition, just set the contained value to `Some(next_state)`.
 /// Note that these transitions can be overridden by other systems:
 /// only the actual value of this resource at the time of [`apply_state_transition`] matters.
-#[derive(Resource, Default, Debug)]
+#[derive(Resource, Default, Clone)]
 #[cfg_attr(
     feature = "bevy_reflect",
     derive(bevy_reflect::Reflect),
     reflect(Resource, Default)
 )]
-pub struct NextState<S: States>(pub Option<S>);
+pub enum NextState<S: States> {
+    /// Do not change the state.
+    #[default]
+    MaintainCurrent,
```

Comment:
I like the naming suggestion, so pulling that in.

As for setting state from a closure or function pointer - this can be useful if you have states that you want to be able to modify an aspect of, but don't care about the rest of the state.

For example, if I want to toggle the pausing of a game, I can pass a function that is transparent to the rest of the state and only modifies the part I want to change. The benefit here is that:
a) you can write a function and pass it in, making it re-usable across your codebase, say if there are multiple ways to trigger a pause
b) you don't need to pre-compute the result, it'll be based on what the state is at the time that apply_state_transition occurs - so if someone manually set the State resource or something it'll still be correct